File: src/common-py/secrets_client_py/secrets_client/_watcher.py
"""
File-system watcher for the secrets-client library.

Uses `watchdog` with the InotifyObserver explicitly — Linux / inotify only.
Fires a callback when a file is moved into place (IN_MOVED_TO), which is the
exact event the CSI driver emits after its atomic rename.

Plain writes (IN_MODIFY / on_modified) are intentionally ignored: the driver
never writes in-place, so a modify event on the secrets mount is either a
staging write to a temp file (irrelevant) or a spurious duplicate (harmless).

DirectoryWatcher places a non-recursive inotify watch on the root directory
and on every real (non-symlink) subdirectory, recursing into new ones as they
appear. Directories whose names start with ".." (e.g. CSI driver versioned
data dirs such as "..data_v2") are deliberately skipped — they accumulate on
long-lived clusters and adding an inotify watch for each would exhaust the
per-process inotify handle limit. The `..data` symlink-swap rotation is
detected at the root level instead.

The watcher runs on a dedicated daemon thread managed by watchdog.
"""

from __future__ import annotations

import threading
from pathlib import Path
from typing import Callable

from watchdog.events import FileSystemEvent, FileSystemEventHandler
from watchdog.observers.inotify import InotifyObserver

RenameCallback = Callable[[Path], None]


class _RenameHandler(FileSystemEventHandler):
    """
    Forwards IN_MOVED_TO events to the appropriate callback.

    File renames go to `on_file`; directory renames that do NOT start with
    ".." go to `on_dir` so DirectoryWatcher can lazily add watches for newly
    appearing subdirectories.
    """

    def __init__(
        self,
        on_file: RenameCallback,
        on_dir: Callable[[Path], None],
    ) -> None:
        super().__init__()
        self._on_file = on_file
        self._on_dir = on_dir

    def on_moved(self, event: FileSystemEvent) -> None:
        dest = Path(str(event.dest_path))
        if event.is_directory:
            # Skip CSI versioned data directories (..data_v2, ..2026_01_01_…).
            if not dest.name.startswith('..'):
                self._on_dir(dest)
        else:
            self._on_file(dest)


class DirectoryWatcher:
    """
    Watches `root_dir` for IN_MOVED_TO events (inotify) and invokes
    `on_rename` with the absolute destination path of the renamed file.

    Places a non-recursive InotifyObserver watch on the root directory and on
    every real (non-symlink) subdirectory, mirroring the TypeScript
    DirectoryWatcher implementation. Directories starting with ".." are never
    watched — see module docstring for rationale.

    Uses InotifyObserver directly — Linux only, no polling fallback.
    """

    def __init__(self, root_dir: Path, on_rename: RenameCallback) -> None:
        self._on_rename = on_rename
        self._observer: InotifyObserver = InotifyObserver()
        self._observer.daemon = True
        self._watched: set[str] = set()
        self._lock = threading.Lock()

        self._watch_dir(root_dir)
        self._observer.start()

    def _watch_dir(self, dir_path: Path) -> None:
        """Add a non-recursive watch for dir_path, then recurse into existing
        real non-'..' subdirectories."""
        key = str(dir_path)
        with self._lock:
            if key in self._watched:
                return
            self._watched.add(key)

        handler = _RenameHandler(self._on_rename, self._on_new_dir)
        self._observer.schedule(handler, key, recursive=False)

        try:
            for entry in dir_path.iterdir():
                if entry.name.startswith('..'):
                    continue
                try:
                    if entry.is_dir() and not entry.is_symlink():
                        self._watch_dir(entry)
                except OSError:
                    pass
        except OSError:
            pass

    def _on_new_dir(self, dir_path: Path) -> None:
        """Lazily watch a directory that was renamed into a watched location."""
        try:
            if dir_path.is_dir() and not dir_path.is_symlink():
                self._watch_dir(dir_path)
        except OSError:
            pass

    def close(self) -> None:
        self._observer.stop()
        self._observer.join(timeout=5)


class Debouncer:
    """
    Collapses rapid calls within `delay_ms` into a single invocation, calling
    the handler once for every *unique* path seen during the window.

    Matches the TypeScript `debounce` helper: accumulating unique paths ensures
    that all rotated files receive a reload call regardless of the order in
    which rename events arrive (e.g. driver writing cert/chain/key in quick
    succession will trigger three separate _handle_rename calls after the
    window, not just the last one).
    """

    def __init__(self, fn: Callable[[Path], None], delay_ms: int) -> None:
        self._fn = fn
        self._delay = delay_ms / 1000.0
        self._timer: threading.Timer | None = None
        self._lock = threading.Lock()
        self._pending: set[Path] = set()

    def __call__(self, file_path: Path) -> None:
        with self._lock:
            self._pending.add(file_path)
            if self._timer is not None:
                self._timer.cancel()
            self._timer = threading.Timer(self._delay, self._flush)
            self._timer.daemon = True
            self._timer.start()

    def _flush(self) -> None:
        with self._lock:
            paths = list(self._pending)
            self._pending.clear()
            self._timer = None
        for p in paths:
            self._fn(p)

    def cancel(self) -> None:
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
            self._pending.clear()

File: src/common-rs/secrets-client/src/watcher.rs
//! File-system watcher for the secrets-client library.
//!
//! Uses the `notify` crate which maps to inotify on Linux and kqueue on macOS.
//! Fires a callback when a file is moved into place (IN_MOVED_TO equivalent),
//! which is the exact event the CSI Secrets Store driver emits after its
//! atomic rename.
//!
//! `DirectoryWatcher` places a non-recursive watch on the root directory and
//! every real (non-symlink) subdirectory. Directories whose names start with
//! ".." (CSI versioned data directories such as "..data_v2") are deliberately
//! skipped to avoid inotify handle exhaustion on long-lived clusters.
//!
//! `Debouncer` collapses rapid calls within `delay_ms` into a single
//! invocation per unique path, matching the TypeScript and Python debounce
//! helpers.

use std::{
    collections::HashSet,
    fs,
    path::{Path, PathBuf},
    sync::{
        atomic::{AtomicBool, Ordering},
        mpsc, Arc,
    },
    thread,
    time::Duration,
};

use notify::{
    event::{EventKind, ModifyKind, RenameMode},
    Config, Event, RecommendedWatcher, RecursiveMode, Watcher,
};

pub type RenameCallback = Arc<dyn Fn(PathBuf) + Send + Sync + 'static>;

// Internal command sent to the DirectoryWatcher worker thread.
enum Cmd {
    Watch(PathBuf),
    Shutdown,
}

/// Watches `root` for rename/create events using a non-recursive inotify watch
/// on the root and every real (non-symlink, non-`..`-prefixed) subdirectory.
/// Mirrors the Go/TypeScript/Python `DirectoryWatcher` implementations.
///
/// When a new real subdirectory appears, the watcher automatically adds it to
/// the watch set. CSI versioned-data directories (`..data_v2`, etc.) are
/// deliberately skipped to avoid inotify fd exhaustion.
///
/// The underlying `RecommendedWatcher` is owned by a background worker thread.
/// Dropping `DirectoryWatcher` sends a `Shutdown` command; the worker exits
/// and drops the watcher cleanly.
pub struct DirectoryWatcher {
    cmd_tx: mpsc::Sender<Cmd>,
}

impl DirectoryWatcher {
    pub fn new(root: PathBuf, on_rename: RenameCallback) -> Result<Self, notify::Error> {
        let (cmd_tx, cmd_rx) = mpsc::channel::<Cmd>();
        let cmd_tx_for_handler = cmd_tx.clone();
        let on_rename_clone = Arc::clone(&on_rename);

        // Build the watcher and register initial watches synchronously so
        // callers can start writing files immediately after new() returns.
        let mut watcher = RecommendedWatcher::new(
            move |res: Result<Event, notify::Error>| {
                if let Ok(event) = res {
                    Self::handle_event(&event, &on_rename_clone, &cmd_tx_for_handler);
                }
            },
            Config::default(),
        )?;

        // Initial non-recursive walk happens HERE (calling thread) to guarantee
        // watches are active before new() returns, avoiding a race with callers
        // that immediately rename files after construction.
        walk_and_watch(&mut watcher, &root);

        thread::spawn(move || {
            let mut w = watcher;
            while let Ok(Cmd::Watch(dir)) = cmd_rx.recv() {
                walk_and_watch(&mut w, &dir);
            }
            // `w` drops here, closing all inotify handles.
        });

        Ok(Self { cmd_tx })
    }

    fn handle_event(event: &Event, on_rename: &RenameCallback, cmd_tx: &mpsc::Sender<Cmd>) {
        let is_relevant = matches!(
            event.kind,
            EventKind::Create(_)
                | EventKind::Remove(_)
                | EventKind::Modify(ModifyKind::Name(
                    RenameMode::To | RenameMode::Both | RenameMode::Any
                ))
        );
        if !is_relevant {
            return;
        }
        for path in &event.paths {
            // For Create/Rename: if a real non-".." subdirectory appeared, queue it.
            // Remove events skip this check (the dir no longer exists).
            if !matches!(event.kind, EventKind::Remove(_)) {
                let base = path
                    .file_name()
                    .map(|n| n.to_string_lossy().into_owned())
                    .unwrap_or_default();
                if !base.starts_with("..") {
                    if let Ok(meta) = fs::symlink_metadata(path) {
                        if meta.is_dir() && !meta.file_type().is_symlink() {
                            let _ = cmd_tx.send(Cmd::Watch(path.clone()));
                        }
                    }
                }
            }
            on_rename(path.clone());
        }
    }
}

impl Drop for DirectoryWatcher {
    fn drop(&mut self) {
        // Signal the worker to exit. The worker drops the watcher on exit,
        // releasing all inotify handles.
        let _ = self.cmd_tx.send(Cmd::Shutdown);
    }
}

/// Adds a non-recursive watch on `dir` and recurses into real (non-symlink,
/// non-`..`-prefixed) subdirectories. Errors (e.g. dir already watched or
/// no longer exists) are silently ignored to keep the watcher idempotent.
fn walk_and_watch(watcher: &mut RecommendedWatcher, dir: &Path) {
    let _ = watcher.watch(dir, RecursiveMode::NonRecursive);
    let Ok(entries) = fs::read_dir(dir) else {
        return;
    };
    for entry in entries.flatten() {
        let name = entry.file_name();
        if name.to_string_lossy().starts_with("..") {
            continue;
        }
        if let Ok(meta) = fs::symlink_metadata(entry.path()) {
            if meta.is_dir() && !meta.file_type().is_symlink() {
                walk_and_watch(watcher, &entry.path());
            }
        }
    }
}

/// Collapses rapid calls within `delay_ms` into a single invocation for each
/// unique path seen during the quiet window. A single background thread is
/// used — no per-call thread spawning.
///
/// The background thread blocks until a path arrives, then accumulates paths
/// until `delay_ms` of silence, then flushes. This matches the
/// TypeScript/Python debounce helpers.
pub struct Debouncer {
    tx: std::sync::Mutex<Option<mpsc::Sender<PathBuf>>>,
    cancelled: Arc<AtomicBool>,
}

impl Debouncer {
    pub fn new(fn_ptr: impl Fn(PathBuf) + Send + Sync + 'static, delay_ms: u64) -> Self {
        let (tx, rx) = mpsc::channel::<PathBuf>();
        let delay = Duration::from_millis(delay_ms);

        let cancelled = Arc::new(AtomicBool::new(false));
        let cancelled_worker = Arc::clone(&cancelled);

        thread::spawn(move || {
            let mut pending = HashSet::<PathBuf>::new();
            loop {
                // Block until at least one path arrives.
                match rx.recv() {
                    Err(_) => break,
                    Ok(path) => {
                        pending.insert(path);
                    }
                }
                // Drain additional paths arriving within the debounce window.
                loop {
                    match rx.recv_timeout(delay) {
                        Ok(path) => {
                            pending.insert(path);
                        }
                        Err(mpsc::RecvTimeoutError::Timeout) => break,
                        Err(mpsc::RecvTimeoutError::Disconnected) => {
                            // Only flush pending paths if the debouncer was not
                            // cancelled — cancel() is intended to discard pending
                            // work (matching the TS cancel() semantics).
                            if !cancelled_worker.load(Ordering::Acquire) {
                                for p in pending.drain() {
                                    fn_ptr(p);
                                }
                            }
                            return;
                        }
                    }
                }
                for p in pending.drain() {
                    fn_ptr(p);
                }
            }
        });

        Self {
            tx: std::sync::Mutex::new(Some(tx)),
            cancelled,
        }
    }

    pub fn call(&self, path: PathBuf) {
        if let Some(tx) = self.tx.lock().unwrap().as_ref() {
            let _ = tx.send(path);
        }
    }

    /// Discards any pending paths and stops the worker thread.
    /// Matches the TypeScript `cancel()` semantics: pending work is dropped,
    /// not flushed.
    pub fn cancel(&self) {
        self.cancelled.store(true, Ordering::Release);
        *self.tx.lock().unwrap() = None;
    }
}
